### ISSUE TITLE
Masking out areas external to NY State

### ISSUE THREAD

**Issue by colinbeier** [caryfitzhugh/NYCCSC-map-viewer#117]
NYSERDA / NESCAUM has asked if its possible to only show the NY State portion for layers that extend beyond its boundaries.  I don't think this is request, but more of a question.  ]

A simple solution would be to create a polygon that 'grays out' everything outside of NYS within a certain area, but this can be problematic when zooming far out.  Steve has done this before (for our GSW project). But it does not really mask the layer(s), just makes them harder to see.  So it does not really work all that well.  

Just wanted to get some thoughts on this really quickly so we can respond to the comment.  

Here is the original comment / question: Map layers – for the ones that encompass more than New York, is it possible to create a toggle to see NY only

**Comment by caryfitzhugh** [comment]
I'm sure we could create a layer which would overlay NY state, and just be transparent in the middle and grey on the outside. So all you would see is NY.

To find a way to clip each layer (but not the base maps) - I think would be much harder. 
@sasignell may have an opinion as well..

**Comment by colinbeier** [comment]
Yep, that is pretty much what I was describing in my initial post.  That's not a great option, but it is easy, and we could test it with NYSERDA.  Other options / approaches will be much more challenging.

**Comment by sasignell** [comment]
agreed.  Want me to make a NY mask?

On Fri, Jan 15, 2016 at 11:03 AM, colinbeier <notifications@github.com>
wrote:

> Yep, that is pretty much what I was describing in my initial post. That's
> not a great option, but it is easy, and we could test it with NYSERDA.
> Other options / approaches will be much more challenging.
>
> On Jan 15, 2016, at 10:56 AM, Cary FitzHugh wrote:
>
> > I'm sure we could create a layer which would overlay NY state, and just
> be transparent in the middle and grey on the outside. So all you would see
> is NY.
> >
> > To find a way to clip each layer (but not the base maps) - I think would
> be much harder.
> > @sasignell may have an opinion as well..
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-172001093>
> .
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by colinbeier** [comment]
Sure, but no hurry... not for this release, of course :)

Colin Beier
SUNY ESF

On Jan 15, 2016 2:56 PM, Steve Signell <notifications@github.com> wrote:
agreed. Want me to make a NY mask?

On Fri, Jan 15, 2016 at 11:03 AM, colinbeier <notifications@github.com>
wrote:

> Yep, that is pretty much what I was describing in my initial post. That's
> not a great option, but it is easy, and we could test it with NYSERDA.
> Other options / approaches will be much more challenging.
>
> On Jan 15, 2016, at 10:56 AM, Cary FitzHugh wrote:
>
> > I'm sure we could create a layer which would overlay NY state, and just
> be transparent in the middle and grey on the outside. So all you would see
> is NY.
> >
> > To find a way to clip each layer (but not the base maps) - I think would
> be much harder.
> > @sasignell may have an opinion as well..
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-172001093>
> .
>



--
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by caryfitzhugh** [comment]
@sasignell @colinbeier -- Masking... we going to do the simple one?
I think I need a NY state geojson mask of some kind.

**Comment by colinbeier** [comment]
Visually would the mask just gray out any layers below it? How far would it extend in space?

Colin Beier
SUNY ESF

On Feb 8, 2016 4:17 PM, Cary FitzHugh <notifications@github.com> wrote:

@sasignell<https://github.com/sasignell> @colinbeier<https://github.com/colinbeier> -- Masking... we going to do the simple one?
I think I need a NY state geojson mask of some kind.

**Comment by caryfitzhugh** [comment]
The easiest would just be 'forever'.   Right?

**Comment by sasignell** [comment]
I'll try to do this tonight

**Comment by sasignell** [comment]
Here's a geojson-- not sure how you want to consume it.  I can makea WMS layer too
[ny_mask.zip](https://github.com/caryfitzhugh/NYCCSC-map-viewer/files/122600/ny_mask.zip)

**Comment by caryfitzhugh** [comment]
I"ve deployed onto gh-pages.  the mask.

**Comment by colinbeier** [comment]
Mask looks great, but layers are drawing on top of it, even if I move the mask to top.

**Comment by sasignell** [comment]
Hey is this supposed to be on by default?

**Comment by caryfitzhugh** [comment]
HA. of course. that needs to go on top.  good catch.

I didn't think it was on by default... @colinbier ?

**Comment by colinbeier** [comment]
Not sure... good question, I will see what Lisa et al. think.

**Comment by colinbeier** [comment]
i'll check about the NYS mask default during our project call tomorrow.

**Comment by sasignell** [comment]
I can send you the quote I sent to Topsham, ME.

Here are some other sites that give quotes

https://apollomapping.com/web-mapping




On Thu, Feb 11, 2016 at 11:01 AM, colinbeier <notifications@github.com>
wrote:

> i'll check about the NYS mask default during our project call tomorrow.
>
> On Feb 11, 2016, at 10:56 AM, Cary FitzHugh wrote:
>
> > HA. of course. that needs to go on top. good catch.
> >
> > I didn't think it was on by default... @colinbier ?
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-182934406>
> .
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by sasignell** [comment]
https://nordpil.com/

http://www.creativeforce.com/custom-maps-2-a-lpp/?gclid=CMDHs7SL8MoCFYpZhgodCpkI5g

On Thu, Feb 11, 2016 at 11:05 AM, Steve Signell <steve@frontierspatial.com>
wrote:

> I can send you the quote I sent to Topsham, ME.
>
> Here are some other sites that give quotes
>
> https://apollomapping.com/web-mapping
>
>
>
>
> On Thu, Feb 11, 2016 at 11:01 AM, colinbeier <notifications@github.com>
> wrote:
>
>> i'll check about the NYS mask default during our project call tomorrow.
>>
>> On Feb 11, 2016, at 10:56 AM, Cary FitzHugh wrote:
>>
>> > HA. of course. that needs to go on top. good catch.
>> >
>> > I didn't think it was on by default... @colinbier ?
>> > —
>> > Reply to this email directly or view it on GitHub.
>> >
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-182934406>
>> .
>>
>
>
>
> --
> Steve Signell
> Frontier Spatial, L.L.C.
> 1062 Dean St.
> Schenectady, NY 12309
> 518-232-1008
> frontierspatial.com
> steve@frontierspatial.com
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by colinbeier** [comment]
@stevesignell you replied to the wrong email!  might want to delete your last two comments on this issue.

**Comment by colinbeier** [comment]
I'm going to ask Lisa about the default, but for now can we get the mask working correctly?  For example the ClimAID layers are still drawing above it - or something is going on that's weird.  


[mask climaid.pdf](https://github.com/caryfitzhugh/NYCCSC-map-viewer/files/131035/mask.climaid.pdf)

**Comment by caryfitzhugh** [comment]
yes. this just needs to be twiddled in z-index. I'll try to do it this
week. Planning on a release at the end of the week with most lf these
things done.

**Comment by sasignell** [comment]
I can probably do this too...I'm going to work on these this afternoon

**Comment by caryfitzhugh** [comment]
For that one steve, in the render function, it usually takes in a z-index.
I think that layer should ignore the z-index coming in, and set it to
something even higher.  I think somewhere in the code (maybe
contorl_panel.js) there are index values defined.  You could add a new one
for "mask"...

**Comment by sasignell** [comment]
So just to see what it looks like, I added the mask in as a permanent layer
that's always on top and took it out of the layer tree.  I kind of like it
and it also says to other states maybe they'd like to pay to get their
state turned on. Let me know what you think.

On Mon, Feb 15, 2016 at 11:17 AM, Cary FitzHugh <notifications@github.com>
wrote:

> For that one steve, in the render function, it usually takes in a z-index.
> I think that layer should ignore the z-index coming in, and set it to
> something even higher. I think somewhere in the code (maybe
> contorl_panel.js) there are index values defined. You could add a new one
> for "mask"...
>
> On Mon, Feb 15, 2016 at 11:15 AM, Steve Signell <notifications@github.com>
> wrote:
>
> > I can probably do this too...I'm going to work on these this afternoon
> > On Feb 15, 2016 11:14 AM, "Cary FitzHugh" <notifications@github.com>
> > wrote:
> >
> > > yes. this just needs to be twiddled in z-index. I'll try to do it this
> > > week. Planning on a release at the end of the week with most lf these
> > > things done.
> > >
> > > On Mon, Feb 15, 2016 at 10:53 AM, colinbeier <notifications@github.com
> >
> > > wrote:
> > >
> > > > I'm going to ask Lisa about the default, but for now can we get the
> > mask
> > > > working correctly? For example the ClimAID layers are still drawing
> > above
> > > > it - or something is going on that's weird.
> > > >
> > > > mask climaid.pdf
> > > > <
> > >
> >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/files/131035/mask.climaid.pdf
> > > >
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <
> > >
> >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184268073
> > > >
> > > > .
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184277109
> > >
> > > .
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184277558
> >
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184278044>
> .
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by colinbeier** [comment]
assume this is on your dev page Steve?  i looked at http://52.2.5.122/NYCCSC-map-viewer/src/map_viewer.html

and the mask works, but map cannot load up any of the ClimAID layers to see how it looks outside of NYS.   i am waiting on an answer about the mask from Lisa et al. I think they want to see it demo-ed before they decide.  I told her next release (this week)... I think the best way to implement for this release is to have it either be in the layer tree, or give it a separate button on the map (if possible) like the navigation and print buttons, but in the upper left corner.   the layer tree option is easier and allows user to adjust opacity... but a button is always coolio.  let's hold off on default for now.

**Comment by sasignell** [comment]
For some reason all the ACIS data is throwing errors when served from that
machine:

MLHttpRequest cannot load
https://d3k0ernyjb0bdf.cloudfront.net/geoserver/www/counties.geojson.json.
The 'Access-Control-Allow-Origin' header has a value 'http://localhost'
that is not equal to the supplied origin. Origin 'http://52.2.5.122' is
therefore not allowed access.

It works on my local machine.



On Mon, Feb 15, 2016 at 7:49 PM, colinbeier <notifications@github.com>
wrote:

> assume this is on your dev page Steve? i looked at
> http://52.2.5.122/NYCCSC-map-viewer/src/map_viewer.html
>
> and the mask works, but map cannot load up any of the ClimAID layers to
> see how it looks outside of NYS. i am waiting on an answer about the mask
> from Lisa et al. I think they want to see it demo-ed before they decide. I
> told her next release (this week)... I think the best way to implement for
> this release is to have it either be in the layer tree, or give it a
> separate button on the map (if possible) like the navigation and print
> buttons, but in the upper left corner. the layer tree option is easier and
> allows user to adjust opacity... but a button is always coolio. let's hold
> off on default for now.
>
> On Feb 15, 2016, at 3:34 PM, Steve Signell wrote:
>
> > So just to see what it looks like, I added the mask in as a permanent
> layer
> > that's always on top and took it out of the layer tree. I kind of like it
> > and it also says to other states maybe they'd like to pay to get their
> > state turned on. Let me know what you think.
> >
> > On Mon, Feb 15, 2016 at 11:17 AM, Cary FitzHugh <
> notifications@github.com>
> > wrote:
> >
> > > For that one steve, in the render function, it usually takes in a
> z-index.
> > > I think that layer should ignore the z-index coming in, and set it to
> > > something even higher. I think somewhere in the code (maybe
> > > contorl_panel.js) there are index values defined. You could add a new
> one
> > > for "mask"...
> > >
> > > On Mon, Feb 15, 2016 at 11:15 AM, Steve Signell <
> notifications@github.com>
> > > wrote:
> > >
> > > > I can probably do this too...I'm going to work on these this
> afternoon
> > > > On Feb 15, 2016 11:14 AM, "Cary FitzHugh" <notifications@github.com>
> > > > wrote:
> > > >
> > > > > yes. this just needs to be twiddled in z-index. I'll try to do it
> this
> > > > > week. Planning on a release at the end of the week with most lf
> these
> > > > > things done.
> > > > >
> > > > > On Mon, Feb 15, 2016 at 10:53 AM, colinbeier <
> notifications@github.com
> > > >
> > > > > wrote:
> > > > >
> > > > > > I'm going to ask Lisa about the default, but for now can we get
> the
> > > > mask
> > > > > > working correctly? For example the ClimAID layers are still
> drawing
> > > > above
> > > > > > it - or something is going on that's weird.
> > > > > >
> > > > > > mask climaid.pdf
> > > > > > <
> > > > >
> > > >
> > >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/files/131035/mask.climaid.pdf
> > > > > >
> > > > > >
> > > > > > —
> > > > > > Reply to this email directly or view it on GitHub
> > > > > > <
> > > > >
> > > >
> > >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184268073
> > > > > >
> > > > > > .
> > > > > >
> > > > >
> > > > > —
> > > > > Reply to this email directly or view it on GitHub
> > > > > <
> > > >
> > >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184277109
> > > > >
> > > > > .
> > > >
> > > > >
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <
> > >
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184277558
> > > >
> > > > .
> > >
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184278044
> >
> > > .
> > >
> >
> >
> >
> > --
> > Steve Signell
> > Frontier Spatial, L.L.C.
> > 1062 Dean St.
> > Schenectady, NY 12309
> > 518-232-1008
> > frontierspatial.com
> > steve@frontierspatial.com
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184455736>
> .
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by caryfitzhugh** [comment]
I can look at this ... can you assign it to me?

**Comment by caryfitzhugh** [comment]
@sasignell I found it frustrating with the auto mask - I wanted to twiddle opacity.  But that might just be me.  And the watersheds aren't on the NY boundaries, so they got cut off as well.  I couldn't see how you had added it, so I couldn't put it back in the layer tree. oh well.  

I think this task should just deal with the layer tree, and the bug above is in a new issue #143

**Comment by sasignell** [comment]
I'll take it out and put it back in the layer tree.. give me a minute

On Tue, Feb 16, 2016 at 3:25 PM, Cary FitzHugh <notifications@github.com>
wrote:

> @sasignell <https://github.com/sasignell> I found it frustrating with the
> auto mask - I wanted to twiddle opacity. But that might just be me. And the
> watersheds aren't on the NY boundaries, so they got cut off as well. I
> couldn't see how you had added it, so I couldn't put it back in the layer
> tree. oh well.
>
> I think this task should just deal with the layer tree, and the bug above
> is in a new issue #143
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/143>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184862808>
> .
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by sasignell** [comment]
OK its back

On Tue, Feb 16, 2016 at 3:26 PM, Steve Signell <steve@frontierspatial.com>
wrote:

> I'll take it out and put it back in the layer tree.. give me a minute
>
> On Tue, Feb 16, 2016 at 3:25 PM, Cary FitzHugh <notifications@github.com>
> wrote:
>
>> @sasignell <https://github.com/sasignell> I found it frustrating with
>> the auto mask - I wanted to twiddle opacity. But that might just be me. And
>> the watersheds aren't on the NY boundaries, so they got cut off as well. I
>> couldn't see how you had added it, so I couldn't put it back in the layer
>> tree. oh well.
>>
>> I think this task should just deal with the layer tree, and the bug above
>> is in a new issue #143
>> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/143>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/caryfitzhugh/NYCCSC-map-viewer/issues/117#issuecomment-184862808>
>> .
>>
>
>
>
> --
> Steve Signell
> Frontier Spatial, L.L.C.
> 1062 Dean St.
> Schenectady, NY 12309
> 518-232-1008
> frontierspatial.com
> steve@frontierspatial.com
>



-- 
Steve Signell
Frontier Spatial, L.L.C.
1062 Dean St.
Schenectady, NY 12309
518-232-1008
frontierspatial.com
steve@frontierspatial.com

**Comment by colinbeier** [comment]
just a heads up, layers are still drawing on top of the mask, not beneath it (on the gh page). the mask shows up at the top of the legends tab, but even at 100% opacity, other layers are still visible outside of NYS. is this fixable?

**Comment by sasignell** [comment]
Yes I just removed the permanent mask and put it back in the layer tree.  I tried to fix this before, but couldn't figure out what was going on with the layer ordering.  I think Cary will have to do this one.

**Comment by caryfitzhugh** [comment]
@colinbeier  I have fixed the mask, so now all the layers render underneath.  Do you want the mask to cover over points as well? I have found (at least this case) where there are a few dam(n) markers which are outside the NY boundaries.  I can't decide which I think is better.  For now I'm putting it as the mask will *not* cover over point markers.  I will deploy this and then if you want to change it, it is a 5s change to make it cover everything.

**Comment by caryfitzhugh** [comment]
@sasignell I had to make it an SVG layer, so it would respect the z-index...  I think they fix this silliness in Leaflet 1.0 (so that all layers can have a z-index).

**Comment by colinbeier** [comment]
guys, can we add the option of a white mask as well as the grey/black one? i just think that will lend itself better to printing and use of the map images in presentations.  hope this is easy peasy.
